src/advertisement/mod.rs

```diff
@@ -0,0 +1,189 @@
+use super::*;
+use core::time::Duration;
+use enr::{CombinedKey, Enr};
+use futures::prelude::*;
+use std::{
+    collections::{HashMap, VecDeque},
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::time::{sleep, Instant, Sleep};
+use tracing::{debug, error};
+
+mod test;
+pub mod ticket;
+
+pub type Topic = [u8; 32];
+
+/// An ad we are adevrtising for another node
+#[derive(Debug)]
+pub struct Ad {
+    node_record: Enr<CombinedKey>,
+    insert_time: Instant,
+}
+
+impl Ad {
+    fn new(node_record: Enr<CombinedKey>, insert_time: Instant) -> Self {
+        Ad {
+            node_record,
+            insert_time,
+        }
+    }
+}
+
+impl PartialEq for Ad {
+    fn eq(&self, other: &Self) -> bool {
+        self.node_record == other.node_record
+    }
+}
+pub struct Ads {
+    expirations: VecDeque<(Pin<Box<Sleep>>, Topic)>,
+    ads: HashMap<Topic, VecDeque<Ad>>,
+    total_ads: i32,
+    ad_lifetime: Duration,
+    max_ads_per_topic: usize,
+    max_ads: i32,
+}
+
+impl Ads {
+    pub fn new(ad_lifetime: Duration, max_ads_per_topic: usize, max_ads: i32) -> Self {
+        Ads {
+            expirations: VecDeque::new(),
+            ads: HashMap::new(),
+            total_ads: 0,
+            ad_lifetime,
+            max_ads_per_topic,
+            max_ads,
+        }
+    }
```

Comment:
I set max_ads_per_topic and max_ads to default values now if the check fails. What action do you think we should take if the check fails? These parameters will eventually be set in the Discv5Config, so maybe the check isn't necessary here after all? @divagant-martian

Reply:
locally (as in wrt the Ads structure) I would return here a `Result<Self, &'static str>`
Then, when discovery starts we can `expect` on it. Discovery's strategy seems to be to panic on wrong configuration so that's what we would do following the rest of the code. Since the discovery config can be created without the need of the builder, the checks on the builder are not enough to guarantee things are right here